src/main/java/com/mongodb/kafka/connect/source/schema/AvroSchema.java

```diff
@@ -0,0 +1,275 @@
+        }
+        break;
+      case ARRAY:
+        validateAvroSchema(avroSchema.getElementType(), fieldPath, recordList);
+        break;
+      case MAP:
+        validateAvroSchema(avroSchema.getValueType(), fieldPath, recordList);
+        break;
+      case UNION:
+        if (avroSchema.getTypes().size() != 2
+            || avroSchema.getTypes().stream()
+                .noneMatch(s -> s.getType() == org.apache.avro.Schema.Type.NULL)) {
+          throw createConnectException(
+              "Union Schemas are not supported, unless one value is "
+                  + "null to represent an optional value.",
+              fieldPath);
+        }
+        break;
+      case FIXED:
+        throw createConnectException(
+            format(
+                "Unsupported Avro schema type: '%s'. The connector will not validate the length. "
+                    + "Use bytes instead.",
+                avroSchema.getType()),
+            fieldPath);
+      case ENUM:
+        throw createConnectException(
+            format(
+                "Unsupported Avro schema type: '%s'. The connector will not validate the values. "
+                    + "Use string instead.",
+                avroSchema.getType()),
+            fieldPath);
+      case STRING:
+      case BYTES:
+      case INT:
+      case LONG:
+      case FLOAT:
+      case DOUBLE:
+      case BOOLEAN:
+        return;
+      case NULL:
+      default:
+        throw createConnectException(
+            format("Unsupported Avro schema type: '%s'.", avroSchema.getType()), fieldPath);
+    }
+  }
+
+  public static Schema fromJson(final String jsonSchema) {
+    org.apache.avro.Schema parsedSchema = validateJsonSchema(jsonSchema);
+    return createSchema(parsedSchema);
+  }
+
+  static org.apache.avro.Schema parseSchema(final String jsonSchema) {
+    try {
+      return new Parser().parse(jsonSchema);
+    } catch (Exception e) {
+      throw new ConnectException(format("Invalid Avro schema. %s\n%s", e.getMessage(), jsonSchema));
+    }
+  }
+
+  static Schema createSchema(final org.apache.avro.Schema avroSchema) {
+    return createSchema(avroSchema, false, new Context());
+  }
+
+  static Schema createSchema(
+      final org.apache.avro.Schema avroSchema, final boolean isOptional, final Context context) {
+    return createSchema(avroSchema, isOptional, null, context);
+  }
+
+  static Schema createSchema(
+      final org.apache.avro.Schema avroSchema,
+      final boolean isOptional,
+      final Object defaultValue,
+      final Context context) {
+    SchemaBuilder builder;
+    switch (avroSchema.getType()) {
+      case RECORD:
+        SchemaBuilder structBuilder = SchemaBuilder.struct();
+        context.schemaCache.put(avroSchema, structBuilder);
+        structBuilder.name(avroSchema.getName());
+        avroSchema
+            .getFields()
+            .forEach(
+                f -> {
+                  if (context.schemaCache.containsKey(f.schema())) {
```

Comment:
There don't appear to be any unit tests where we enter this conditional

Reply:
Added